### PR TITLE
add next.config.js validation to turbopack

### PR DESCRIPTION
**Pull request by ForsakenHarmony**

### CODE REVIEW

**Comment by ForsakenHarmony** [comment]
Current dependencies on/for this PR:
* main
  * **PR #3105** <a href="https://app.graphite.dev/github/pr/vercel/turbo/3105" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/vercel/turbo/3105?utm_source=stack-comment).

File: crates/next-core/js/src/entry/config/next.js [hunk 1]
```diff
@@ -3,8 +3,16 @@ import { PHASE_DEVELOPMENT_SERVER } from "next/dist/shared/lib/constants";
 
 const loadNextConfig = async () => {
   const nextConfig = await loadConfig(PHASE_DEVELOPMENT_SERVER, process.cwd());
-  nextConfig.rewrites = await nextConfig.rewrites?.();
+
+  nextConfig.generateBuildId = await nextConfig.generateBuildId?.();
+  nextConfig.headers = await nextConfig.headers?.();
```
Comment (jridgewell) [inline crates/next-core/js/src/entry/config/next.js:8]: the `loadCustomRoutes` from `next/dist/lib/load-custom-routes` will perform some validation and loading for `headers`/`redirects`/`rewrites`.

File: crates/next-core/src/next_config.rs [hunk 2]
```diff
@@ -29,29 +31,212 @@ use turbopack_node::{
...
+struct DevIndicatorsConfig {
+    build_activity: bool,
+    build_activity_position: BuildActivityPositions,
 }
 
-#[derive(Clone, Debug, Ord, PartialOrd, PartialEq, Eq, Serialize, Deserialize, TraceRawVcs)]
+#[derive(Clone, Debug, PartialEq, Serialize, Deserialize, TraceRawVcs)]
+#[serde(rename_all = "camelCase")]
+struct OnDemandEntriesConfig {
+    max_inactive_age: f64,
```
Comment (jridgewell) [inline crates/next-core/src/next_config.rs:82]: Are these floats or u64s?

**Review by jridgewell — APPROVED** [review]

**Comment by jridgewell** [comment]
I admittedly didn't fully review every line.

**Comment by alexkirsz** [comment]
I'm also validating next.config.js in https://github.com/vercel/turbo/pull/3071 so that will likely cause a conflict.

File: crates/next-core/src/next_config.rs (hunk 2 above)
Comment (ForsakenHarmony) [reply crates/next-core/src/next_config.rs:82]: yeah, ideally these would be `u64`, but for JS compat I've put them in here as `f64` for now

File: crates/next-core/Cargo.toml [hunk 3]
```diff
@@ -11,9 +11,11 @@ bench = false
 [dependencies]
 anyhow = "1.0.47"
 auto-hash-map = { path = "../auto-hash-map" }
+crossterm = "0.25"
```
[the remark this answers is not in the document]
Comment (ForsakenHarmony) [reply crates/next-core/Cargo.toml:14]: I think there are a lot of dependencies that are used across multiple crates but are not at the workspace level, `anyhow` above is a good example

Should we move all of those to the workspace?

File: crates/next-core/src/next_config.rs [hunk 4]
```diff
@@ -119,52 +302,186 @@ pub struct RemotePattern {
...
+            fetch_cache: Some(false),
+            font_loaders: None,
+            fully_specified: Some(false),
+            gzip_size: Some(true),
+            incremental_cache_handler_path: None,
+            isr_flush_to_disk: Some(true),
+            // default to 50MB limit
+            isr_memory_cache_size: Some(50.0 * 1024.0 * 1024.0),
+            // default to 128KB
+            large_page_data_bytes: Some(128.0 * 1000.0),
```
[the remark this answers is not in the document]
Comment (ForsakenHarmony) [reply crates/next-core/src/next_config.rs:419]: It was, yes
https://github.com/vercel/next.js/blob/canary/packages/next/src/server/config-shared.ts#L607

File: crates/next-core/src/next_config.rs [hunk 5]
```diff
@@ -263,11 +580,190 @@ pub async fn load_next_config(execution_context: ExecutionContextVc) -> Result<N
...
+const SUPPORTED_OPTIONS: &[&'static str] = &[
+    "images",
+    "reactStrictMode",
+    "configFileName",
+    "swcMinify",
+    "experimental.appDir",
+    "experimental.serverComponentsExternalPackages",
+];
+
+pub async fn validate_next_config(
```
[the remark this answers is not in the document]
Comment (ForsakenHarmony) [reply crates/next-core/src/next_config.rs:657]: Nope 👍

File: crates/next-core/Cargo.toml (hunk 3 above)
[the remark this answers is not in the document]
Comment (alexkirsz) [reply crates/next-core/Cargo.toml:14]: Yes. I created [WEB-174](https://linear.app/vercel/issue/WEB-174/normalize-common-dependencies-versions-across-the-workspace) a while back to tackle this.

**Comment by ForsakenHarmony** [comment]
closing for now

**Pull request closed without merging**